In [5]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/House_Rent_Dataset.csv")

In [11]:
# Extract the current floor number
df["Floor_Number"] = (
    df["Floor"]
    .str.extract(r"^(\d+)")
    .fillna(0)
    .astype(int)
)

# Extract the total number of floors in the building
df["Total_Floors"] = (
    df["Floor"]
    .str.extract(r"out of (\d+)")
    .fillna(0)
    .astype(int)
)

In [12]:
df[["Floor", "Floor_Number", "Total_Floors"]].head(10)

,Floor,Floor_Number,Total_Floors
0,Ground out of 2,0,2
1,1 out of 3,1,3
2,1 out of 3,1,3
3,1 out of 2,1,2
4,1 out of 2,1,2
5,Ground out of 1,0,1
6,Ground out of 4,0,4
7,1 out of 2,1,2
8,1 out of 2,1,2
9,1 out of 3,1,3


In [13]:
df["Posted On"].unique()

<StringArray>
['18-05-2022', '13-05-2022', '16-05-2022', '04-07-2022', '09-05-2022',
 '29-04-2022', '21-06-2022', '07-06-2022', '20-06-2022', '23-05-2022',
 '14-05-2022', '05-05-2022', '01-06-2022', '17-05-2022', '09-06-2022',
 '02-07-2022', '14-06-2022', '15-06-2022', '28-05-2022', '22-05-2022',
 '18-06-2022', '25-06-2022', '22-06-2022', '21-05-2022', '26-06-2022',
 '16-06-2022', '29-06-2022', '10-05-2022', '12-05-2022', '03-06-2022',
 '31-05-2022', '10-06-2022', '24-06-2022', '30-04-2022', '27-05-2022',
 '06-06-2022', '20-05-2022', '26-05-2022', '03-07-2022', '23-04-2022',
 '19-06-2022', '06-07-2022', '04-06-2022', '30-05-2022', '08-06-2022',
 '03-05-2022', '07-05-2022', '27-06-2022', '15-05-2022', '28-04-2022',
 '28-06-2022', '23-06-2022', '13-06-2022', '04-05-2022', '25-04-2022',
 '06-05-2022', '30-06-2022', '01-07-2022', '17-06-2022', '11-06-2022',
 '11-05-2022', '12-06-2022', '05-06-2022', '25-05-2022', '24-05-2022',
 '19-05-2022', '24-04-2022', '27-04-2022', '08-05-2022', '26-04

In [15]:
df["Posted On"] = pd.to_datetime(
    df["Posted On"],
    format="%d-%m-%Y"
)
df[["Posted On"]].head()

,Posted On
0,2022-05-18
1,2022-05-13
2,2022-05-16
3,2022-07-04
4,2022-05-09


In [17]:
df["Posted_Month"] = df["Posted On"].dt.month
df["Posted_Day"] = df["Posted On"].dt.day
df[["Posted On", "Posted_Month", "Posted_Day"]].head(10)

,Posted On,Posted_Month,Posted_Day
0,2022-05-18,5,18
1,2022-05-13,5,13
2,2022-05-16,5,16
3,2022-07-04,7,4
4,2022-05-09,5,9
5,2022-04-29,4,29
6,2022-06-21,6,21
7,2022-06-21,6,21
8,2022-06-07,6,7
9,2022-06-20,6,20


In [18]:
locality_counts = df["Area Locality"].value_counts()

print("Localities appearing only once:", (locality_counts == 1).sum())
print("Localities appearing 2-5 times:", locality_counts.between(2, 5).sum())

Localities appearing only once: 1465
Localities appearing 2-5 times: 605


In [19]:
categorical_columns = df.select_dtypes(include="object").columns

print(categorical_columns)

Index(['Floor', 'Area Type', 'Area Locality', 'City', 'Furnishing Status',
       'Tenant Preferred', 'Point of Contact'],
      dtype='str')


C:\Users\Seema\AppData\Local\Temp\ipykernel_7436\1123855639.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


In [26]:
print(df.columns.tolist())

['BHK', 'Rent', 'Size', 'Area Type', 'City', 'Furnishing Status', 'Tenant Preferred', 'Bathroom', 'Point of Contact', 'Floor_Number', 'Total_Floors', 'Posted_Month', 'Posted_Day']


In [25]:
df.drop("Posted On", axis=1, inplace=True)

In [27]:
categorical_columns = [
    "Area Type",
    "City",
    "Furnishing Status",
    "Tenant Preferred",
    "Point of Contact"
]

for col in categorical_columns:
    print(f"\n{col}:")
    print(df[col].unique())


Area Type:
<StringArray>
['Super Area', 'Carpet Area', 'Built Area']
Length: 3, dtype: str

City:
<StringArray>
['Kolkata', 'Mumbai', 'Bangalore', 'Delhi', 'Chennai', 'Hyderabad']
Length: 6, dtype: str

Furnishing Status:
<StringArray>
['Unfurnished', 'Semi-Furnished', 'Furnished']
Length: 3, dtype: str

Tenant Preferred:
<StringArray>
['Bachelors/Family', 'Bachelors', 'Family']
Length: 3, dtype: str

Point of Contact:
<StringArray>
['Contact Owner', 'Contact Agent', 'Contact Builder']
Length: 3, dtype: str


In [31]:
print(df.columns.tolist())

['BHK', 'Rent', 'Size', 'Bathroom', 'Floor_Number', 'Total_Floors', 'Posted_Month', 'Posted_Day', 'Area Type_Carpet Area', 'Area Type_Super Area', 'City_Chennai', 'City_Delhi', 'City_Hyderabad', 'City_Kolkata', 'City_Mumbai', 'Furnishing Status_Semi-Furnished', 'Furnishing Status_Unfurnished', 'Tenant Preferred_Bachelors/Family', 'Tenant Preferred_Family', 'Point of Contact_Contact Builder', 'Point of Contact_Contact Owner']


In [40]:
df.dtypes

BHK                                  int64
Rent                                 int64
Size                                 int64
Bathroom                             int64
Floor_Number                         int64
Total_Floors                         int64
Posted_Month                         int64
Posted_Day                           int64
Area Type_Carpet Area                int64
Area Type_Super Area                 int64
City_Chennai                         int64
City_Delhi                           int64
City_Hyderabad                       int64
City_Kolkata                         int64
City_Mumbai                          int64
Furnishing Status_Semi-Furnished     int64
Furnishing Status_Unfurnished        int64
Tenant Preferred_Bachelors/Family    int64
Tenant Preferred_Family              int64
Point of Contact_Contact Builder     int64
Point of Contact_Contact Owner       int64
dtype: object

In [39]:
bool_columns = df.select_dtypes(include="bool").columns

df[bool_columns] = df[bool_columns].astype(int)

In [41]:
X = df.drop("Rent", axis=1)
y = df["Rent"]
print("X shape:", X.shape)
print("y shape:", y.shape)
print(X.columns.tolist())

X shape: (4746, 20)
y shape: (4746,)
['BHK', 'Size', 'Bathroom', 'Floor_Number', 'Total_Floors', 'Posted_Month', 'Posted_Day', 'Area Type_Carpet Area', 'Area Type_Super Area', 'City_Chennai', 'City_Delhi', 'City_Hyderabad', 'City_Kolkata', 'City_Mumbai', 'Furnishing Status_Semi-Furnished', 'Furnishing Status_Unfurnished', 'Tenant Preferred_Bachelors/Family', 'Tenant Preferred_Family', 'Point of Contact_Contact Builder', 'Point of Contact_Contact Owner']


## Feature Engineering Summary

- Extracted `Floor_Number` and `Total_Floors` from the `Floor` column.
- Converted `Posted On` into `Posted_Month` and `Posted_Day`.
- Removed `Area Locality` due to its high cardinality (2,235 unique localities).
- Removed the original `Floor` and `Posted On` columns after extracting useful information.
- Applied one-hot encoding to categorical features such as City, Area Type, Furnishing Status, Tenant Preferred, and Point of Contact.
- Separated the target variable `Rent` from the input features.

The final dataset contains **4,746 records and 20 input features** for rent prediction.